listenbrainz_spark/stats_writer/stats_writer.py

```diff
@@ -0,0 +1,70 @@
+import pika
+import logging
+from listenbrainz_spark import config
+import time
+import json
+import sys
+
+class StatsWriter():
+    def __init__(self):
+        self.unique_ch = None
+        self.ERROR_RETRY_DELAY = 3
+
+    def connect_to_rabbitmq(self):
+        """Creates a RabbitMQ connection object
+        """
+        username = config.RABBITMQ_USERNAME
+        password = config.RABBITMQ_PASSWORD
+        host = config.RABBITMQ_HOST
+        port = config.RABBITMQ_PORT
+        virtual_host = config.RABBITMQ_VHOST
+
+        while True:
+            try:
+                credentials = pika.PlainCredentials(username, password)
+                connection_parameters = pika.ConnectionParameters(
+                    host=host,
+                    port=port,
+                    virtual_host=virtual_host,
+                    credentials=credentials,
+                )
+                self.connection =  pika.BlockingConnection(connection_parameters)
+                break
+            except Exception as err:
+                error_message = "Cannot connect to RabbitMQ: {error}, retrying in {delay} seconds."
+                print(error_message.format(error=str(err), delay=self.ERROR_RETRY_DELAY))
+                time.sleep(self.ERROR_RETRY_DELAY)
+
+    def write(self, data):
+        """Publish data to RabbitMQ
+        """
+        try:
+            self.unique_ch.basic_publish(
+                exchange=config.SPARK_EXCHANGE,
+                routing_key='',
+                body=json.dumps(data),
+                properties=pika.BasicProperties(delivery_mode = 2,),
+            )
+        except pika.exceptions.ConnectionClosed as e:
+            logging.error("Connection to rabbitmq closed while trying to publish: %s" % str(e), exc_info=True)
+        except Exception as e:
+            logging.error("Cannot publish to rabbitmq channel: %s / %s" % (type(e).__name__, str(e)), exc_info=True)
+
+
+    def start(self, data):
+        """Establishes connection to RabbitMQ if not connected
+        """
+        if not config.RABBITMQ_HOST:
+            logging.critical("RabbitMQ service not defined. Sleeping {0} seconds and exiting.".format(self.ERROR_RETRY_DELAY))
+            time.sleep(self.ERROR_RETRY_DELAY)
+            sys.exit(-1)
+
+        if self.unique_ch:
```

Comment:
There exists a case where we might have opened a connection and have a unique_ch object -- however, that object might be stale and the connection to RabbitMQ has been dropped. This means the subsequent write() will fail and right now that error will fail quietly. 

I think rewriting this function to:

0. In an infinite loop:
1. Check to see if we have a connection by checking unique_ch, if not open one by calling connect_to_rabbitmq
2. Attempt the write. If it works, exit the loop.
3. If the write fails, clean up the existing connection (if we can/need) and set unique_ch to None. Then repeat the loop.